### PR TITLE
Update 0912-sort-an-array-medium.md

### PR DESCRIPTION
**Pull request by deepanshu-rawat6**
## Change Summary

Provide summary of changes with issue number if any.

## Checklist

**If you haven't fulfilled the below requirements, your PR won't be reviewed.**

### General 

- [X] Make sure you've read [CONTRIBUTING.md](https://github.com/wingkwong/leetcode-the-hard-way/blob/main/CONTRIBUTING.md) 
- [X] Make sure you've followed the template format
- [X] Make sure you've started the app locally and verified all the content and all links (if applicable) must be accessible correctly
- [X] Make sure you've applied LaTex for variables, formulas and time / space complexity instead of using backticks
- [X] Make sure you've formatted your code well (Code Block is rendered using `<Tabs/>`)
      Below is a C++ example. You can include multiple langs using `<TabItem/>` if necessary
      
````
<Tabs>

<TabItem value="cpp" label="C++">
<SolutionAuthor name="@YOUR_ALIAS"/>

```cpp
// Your code goes here
// Prefer K&R Coding style
```

</TabItem>

</Tabs>
````
- [X] Make sure you've explained your code well (write comments above each line - not inline)
- [X] Make sure you've written your explanation well and it is easy to understand for beginners
- [X] Make sure you've included Time & Space Complexity Analysis
- [X] Make sure you've included a meaningful approach name for your solution. Don't leave it blank. e.g. `## Approach 1: Two Pointers`.

### For Tutorial Articles

- [ ] Make sure you've explained your topic using 1 - 3 LC problems rather than using external problems. Solutions must be taken with explanations as well.
- [ ] Make sure you've provided suggested problems at the end using the given format. See [here](https://raw.githubusercontent.com/wingkwong/leetcode-the-hard-way/main/tutorials/math/number-theory/binary-exponentiation.md) as an example. If the target solution is not available, leave `solutionLink` blank.
- [ ] Make sure you've given credits / references if you use external resources. For an image, give credit under it. Otherwise, add a new section called References at the end (after Suggested Problems).

### CODE REVIEW

**Comment by wingkwong** [comment]
DON'T close the PR and create a new every time you change something. You can just push your changes and they will be shown in the same PR.

ref: #179 #227

**Review by wingkwong — CHANGES_REQUESTED** [review]
1. Need more explanation on your approach and solutions. your audience is beginners. don't expect a few lines can explain everything.
2. Verify it locally first please

image

**Comment by deepanshu-rawat6** [comment]
@wingkwong both detailed explanation and local verification has been made, kindly review these changes

File: solutions/0900-0999/0912-sort-an-array-medium.md [hunk 1]
```diff
@@ -173,4 +173,155 @@ void quickSort(vector<int>& nums, int l, int r) {
...
+
+### Merge Sort with Outplace Merging
+
+In this approach, merging happen in $seperate array$ of size, then it is passed on in the above function calls.
+
+In mergeSort() function:
+    Dividing the array into 2 subarrays(one is $left[]$ and other is $right[]$ ) till their length reaches one.
+
+In merge() function:
+    We merge the two subarrays( $left[]$ and $right[]$ ) into another array $mix[]$ . This merge operation stores elements from
```
Comment (wingkwong) [inline solutions/0900-0999/0912-sort-an-array-medium.md:191]: same as above

File: solutions/0900-0999/0912-sort-an-array-medium.md [hunk 2]
```diff
@@ -173,4 +173,155 @@ void quickSort(vector<int>& nums, int l, int r) {
...
+<SolutionAuthor name="@deepanshu-rawat6"/>
+
+```java
+public static int[] mergeSort(int[] arr) {
+    if (arr.length == 1) {
+        return arr;
+    }
+
+    int mid = arr.length / 2;
+    // coppying and sorting sub-array by division on the basis of mid value
```
Comment (wingkwong) [inline solutions/0900-0999/0912-sort-an-array-medium.md:209]: typo

File: solutions/0900-0999/0912-sort-an-array-medium.md [hunk 3]
```diff
@@ -173,4 +173,155 @@ void quickSort(vector<int>& nums, int l, int r) {
...
+
+    return mix;
+}
+```
+</TabItem>
+</Tabs>
+
+### Merge Sort using In-place Merging
+
+In this approach, merging happens $in-place$, changes are made in the original arrays itself by modifying the reference variables.
```
Comment (wingkwong) [inline solutions/0900-0999/0912-sort-an-array-medium.md:257]: Incorrect usage of LaTex on `$in-place$`. Change it to `in-place`.

**Review by wingkwong — CHANGES_REQUESTED** [review]

**Comment by deepanshu-rawat6** [comment]
@wingkwong I've fixed these changes now

**Comment by deepanshu-rawat6** [comment]
@wingkwong Ma'am can you please review my PR now?

File: solutions/0900-0999/0912-sort-an-array-medium.md [hunk 4]
```diff
@@ -173,4 +173,155 @@ void quickSort(vector<int>& nums, int l, int r) {
...
+In this approach, merging happen in `seperate array` of size, then it is passed on in the above function calls.
+
+In mergeSort() function:
+    Dividing the array into 2 subarrays(one is $left$ and other is $right$) till their length reaches one(because no futher division can take place), then passing left and right parts into merge() function for sorting in correct order.
+
+In merge() function:
+    We merge the two subarrays( $left$ and $right$ ) into another array $mix$ . This merge operation stores elements from 
+    $left$ and $right$ in an ascending order, by comparing elements from individual subarrays.
+    Finally, since one subarray would have greater elements then the other one, so we add those elements seperately.
+    Then, we return the sorted subarray into the above function call.
```
Comment (wingkwong) [inline solutions/0900-0999/0912-sort-an-array-medium.md:194]: please run the app locally to verify. the content is all packed which is quite hard to read.

File: solutions/0900-0999/0912-sort-an-array-medium.md [hunk 5]
```diff
@@ -173,4 +173,155 @@ void quickSort(vector<int>& nums, int l, int r) {
...
+
+In this approach, merging happens `in-place`, changes are made in the original arrays itself by modifying the reference variables. 
+
+In mergeSort() function:
+    Dividing the array into 2 subarrays(by calling mergeSort() function recursively) till their length reaches one.
+
+In mergeInPlace() function:
+    We merge the two subarrays(one from s to mid and another from mid to e) into another array $mix$ . This merge operation stores elements from the two sub-arrays in an ascending order, by comparing elements from individual subarrays.
+    Finally, since one subarray would have greater elements then the other one, so we add those elements seperately.
+    Then, we place the sorted elements from $mix$ into $arr$ , finally changes in $arr$ have been made in-place.
```
Comment (wingkwong) [inline solutions/0900-0999/0912-sort-an-array-medium.md:265]: same as above

File: solutions/0900-0999/0912-sort-an-array-medium.md [hunk 6]
```diff
@@ -173,4 +173,155 @@ void quickSort(vector<int>& nums, int l, int r) {
...
+</Tabs>
+
+## Approach 3: Merge Sort
+
+Standard merge sort algorithm, following a divide and conquer approach, division of sub-array by mid values till sub-array length
+reaches one, then merging parts by either out-place or in-place merging in either ascending or descending order. 
+
+### Merge Sort with Outplace Merging
+
+In this approach, merging happen in `seperate array` of size, then it is passed on in the above function calls.
```
Comment (wingkwong) [inline solutions/0900-0999/0912-sort-an-array-medium.md:185]: remove backticks

File: solutions/0900-0999/0912-sort-an-array-medium.md [hunk 7]
```diff
@@ -173,4 +173,155 @@ void quickSort(vector<int>& nums, int l, int r) {
...
+
+    return mix;
+}
+```
+</TabItem>
+</Tabs>
+
+### Merge Sort using In-place Merging
+
+In this approach, merging happens `in-place`, changes are made in the original arrays itself by modifying the reference variables.
```
Comment (wingkwong) [inline solutions/0900-0999/0912-sort-an-array-medium.md:257]: remove backticks

File: solutions/0900-0999/0912-sort-an-array-medium.md [hunk 8]
```diff
@@ -173,4 +173,155 @@ void quickSort(vector<int>& nums, int l, int r) {
...
+
+    while (j < e) {
+        mix[k] = arr[j];
+        j++;
+        k++;
+    }
+
+    // modifying the original arrays by replacing elements into their correct indices
+    for (int l = 0; l < mix.length; l++) {
+        arr[s+l] = mix[l];
```
Comment (wingkwong) [inline solutions/0900-0999/0912-sort-an-array-medium.md:322]: add space.

```
arr[s + l]
```

File: solutions/0900-0999/0912-sort-an-array-medium.md [hunk 9]
```diff
@@ -173,4 +173,155 @@ void quickSort(vector<int>& nums, int l, int r) {
...
+<SolutionAuthor name="@deepanshu-rawat6"/>
+
+```java
+public void mergeSort(int[] arr, int s, int e) {
+    if (e - s == 1) {
+        return;
+    }
+
+    int mid = (s + e) / 2;
+    // Dividing sub-arrays by mid values till, sub-array length reaches one
```
Comment (wingkwong) [inline solutions/0900-0999/0912-sort-an-array-medium.md:280]: start with lower case

**Review by wingkwong — CHANGES_REQUESTED** [review]

**Comment by deepanshu-rawat6** [comment]
@wingkwong can I add explanation in bullet points?

**Comment by wingkwong** [comment]
Yes. Go ahead. Please check it locally before requesting review.

**Comment by deepanshu-rawat6** [comment]
@wingkwong I have now verified the changes locally

**Pull request closed without merging**